sdk/eventhub/event-hubs/src/eventHubSender.ts

```diff
@@ -508,164 +498,202 @@ export class EventHubSender extends LinkEntity {
+          const err = translate(context!.delivery!.remote_state!.error);
+          log.error(err);
+          reject(err);
+        };
+        onReleased = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event released.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender '${this.name}', ` +
+                `received a release disposition.Hence we are rejecting the promise.`
             );
-            resolve();
-          };
-          onRejected = (context: EventContext) => {
-            removeListeners();
-            log.error(
-              "[%s] Sender '%s', got event rejected.",
-              this._context.connectionId,
-              this.name
+          }
+          log.error(err);
+          reject(err);
+        };
+        onModified = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event modified.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender "${this.name}", ` +
+                `received a modified disposition.Hence we are rejecting the promise.`
             );
-            const err = translate(context!.delivery!.remote_state!.error);
-            log.error(err);
-            reject(err);
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const removeListeners = (): void => {
+          // When `removeListeners` is called on timeout, the sender might be closed and cleared
+          // So, check if it exists, before removing listeners from it.
+          if (abortSignal) {
+            abortSignal.removeEventListener("abort", onAborted);
+          }
+          if (this._sender) {
+            this._sender.removeListener(SenderEvents.rejected, onRejected);
+            this._sender.removeListener(SenderEvents.accepted, onAccepted);
+            this._sender.removeListener(SenderEvents.released, onReleased);
+            this._sender.removeListener(SenderEvents.modified, onModified);
+          }
+        };
+
+        const rejectOnSendError = (err: Error) => {
+          err = translate(err);
+          log.error(
+            "[%s] An error occurred while performing send on %s",
+            this._context.connectionId,
+            this.name,
+            err
+          );
+          reject(err);
+        };
+
+        const actionAfterTimeout = () => {
+          clearTimeout(waitTimer);
```

Comment:
Shouldn't clearTimeout be called inside `removeListeners`?  Right now you're only calling `clearTimeout` inside of the function that gets triggered after the timeout, so it isn't really doing anything.

Reply:
Oh that's right, it should get cleared on all early exits - updated!